### PR TITLE
FEAT: Report generation screen

### PR DESCRIPTION
**Pull request by sutgeorge**
## Changes

- the newly created screen
- fix: the menu dropdown from the upper app bar was not closed after selecting an item
- fix: banner images displayed at the bottom of the HotspotForm and ReportGeneration screens did not fit the whole width of the device screen on the web platform

### iOS

https://github.com/crafting-software/nuca-mobile/assets/32801760/728e706a-bf69-4360-9d8f-0daf3503c2d8

### Android

https://github.com/crafting-software/nuca-mobile/assets/32801760/38eb741b-3a01-4876-95c8-0cdc21b3a632

### Web

https://github.com/crafting-software/nuca-mobile/assets/32801760/7341b475-6939-42da-86f4-94f12651f2fa

### CODE REVIEW

File: screens/ReportGenerationScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+              <View style={styles.screenTitleContainer}>
+                <View style={styles.titleRow}>
+                  <Headline style={styles.screenTitleIcon}>
+                    <View>
+                      <Icon source={'file-document-outline'} size={22} />
+                    </View>
+                  </Headline>
+                  <Caption style={styles.screenTitleLabel}>
+                    Generează raport
+                  </Caption>
```
Comment (mveres) [inline screens/ReportGenerationScreen.tsx:119]: The icon and the caption are not aligned horizontally as seen in the screen recording. Can you adjust this?

File: screens/ReportGenerationScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      <Appbar forDetailScreen />
+      <View style={styles.container}>
+        <ScrollView style={styles.scrollView}>
+          <View
+            style={{
+              alignItems: 'center',
+            }}
+          >
+            <View style={styles.contentContainer}>
+              <View style={styles.screenTitleContainer}>
```
Comment (mveres) [inline screens/ReportGenerationScreen.tsx:110]: these two Views can be merged into one

File: screens/ReportGenerationScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+import { FullScreenActivityIndicator } from '../components/FullScreenActivityIndicator';
+import { useNucaTheme as useTheme } from '../hooks/useNucaTheme';
+import { RootStackParamList } from '../types';
+import { isSmallScreen } from '../utils/helperFunc';
+
+const getStyles = (theme: NucaCustomTheme) =>
+  StyleSheet.create({
+    saveButton: {
+      height: 64,
+      maxWidth: 340,
```
Comment (mveres) [inline screens/ReportGenerationScreen.tsx:17]: These dimensions should be set at project level an reused in all button (or other components)
What's the approach so far? Can you extract these?
Having dimensions hardcoded per component make the app more rigid and not friendly to design changes.

File: screens/ReportGenerationScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  return (
+    <>
+      <Appbar forDetailScreen />
+      <View style={styles.container}>
+        <ScrollView style={styles.scrollView}>
+          <View
+            style={{
+              alignItems: 'center',
+            }}
```
Comment (mveres) [inline screens/ReportGenerationScreen.tsx:107]: why is this style inlined?

File: screens/ReportGenerationScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+) => {
+  const theme = useTheme();
+  const styles = getStyles(theme);
+  const [isInProgress, _setIsInProgress] = useState(false);
+
+  return (
+    <>
+      <Appbar forDetailScreen />
+      <View style={styles.container}>
+        <ScrollView style={styles.scrollView}>
```
Comment (mveres) [inline screens/ReportGenerationScreen.tsx:103]: I noticed in the screens that the background image extends lower than the white background... can you fix this?

**Review by mveres — CHANGES_REQUESTED** [review]

File: screens/ReportGenerationScreen.tsx (hunk 3 above)
Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx:17]: True, I think we can create a re-usable button component with this blue button (the one on the report generation screen is identical to the one on the HotspotForm).

Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx:17]: Perhaps calling it `NucaFormButton` makes sense, since it is only used on form-like screens (HotspotFormScreen.ts, ReportGenerationScreen.ts).

File: screens/ReportGenerationScreen.tsx (hunk 4 above)
Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx]: I mentioned this in a comment above.

TLDR: This view was redundant and the `alignItems` prop was applied on the immediate child of this View (and its value was changed to `stretch`). More details in the walltext written above 😄

File: screens/ReportGenerationScreen.tsx (hunk 5 above)
Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx:88]: Disabling the `ScrollView` solves the issue on mobile, but not on web. I tried to remove the `ScrollView`, but the contents of the screen are not displayed as expected (date picker captions are displayed behind the date picker elements themselves).
Other than that, I tried using the `transform: translate` prop, but that doesn't seem to work at all. Perhaps we could note down this bug as a separate task?

File: screens/ReportGenerationScreen.tsx (hunk 3 above)
Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx]: Anyways, with the same dimensions, the re-usable button component seems to be displayed as expected on all platforms.

**Comment by sutgeorge** [comment]
https://github.com/crafting-software/nuca-mobile/pull/76

File: components/Footer.tsx [hunk 6]
```diff
@@ -20,6 +21,14 @@ const getFooterStyles = (_theme: NucaCustomTheme) =>
       flex: 1,
       width: undefined,
     },
+    zoomedImage: {
+      paddingTop: 12,
+      height: 375,
+      minHeight: 430,
+      flex: 1,
+      width: undefined,
```
Comment (salajandan) [inline components/Footer.tsx:29]: why is the width `undefined` ?

Comment (sutgeorge) [reply components/Footer.tsx:29]: I didn't personally set the width to `undefined`. However, after a closer look, I found out that the reason why the width is set to `undefined` is to ensure that the bottom banner image takes up the width of its parent component i.e. the container. The [answer in this thread](https://w3schools.invisionzone.com/topic/57959-when-an-elements-width-is-undefined-what-determines-it/?do=findComment&comment=318035) explains the issue. 

I tried to see what happens if the `width: undefined` is set to `width: '100%'` and the result is that on some devices selected in the developer console in Chrome, the bottom banner is properly displayed, so the `undefined` setting was totally pointless (setting the width to `undefined` actually makes the bottom banner invisible on some devices, or at least that's what the developer console screen size settings indicate).

Anyway, good catch!

Comment (sutgeorge) [reply components/Footer.tsx]: I just pushed a new commit that covers the width issue that was mentioned.

Changes:

1. the 'width' prop is now set to '100%' instead of 'undefined'
2. the footer image/bottom banner is now positioned absolutely at a distance of 0 units from the bottom of the parent container when the screen is large and positioned at a distance of '-20%' from the bottom of the screen on small devices (this was tested on all devices in the Chrome developer console, from which only "Surface Duo" does not display the banner as expected)
3. the NucaFormButton has a zIndex of 2 and the footer image has a zIndex of -1 => on some devices in the Chrome console, the banner is displayed on top of the form contents
4. the ScrollView is not scrollable and its height is now set to '100%' instead of '1000%'

**Review by mveres — APPROVED** [review]

File: components/Footer.tsx (hunk 6 above)
Comment (sutgeorge) [reply components/Footer.tsx]: The Surface Duo issue is now fixed.

https://github.com/crafting-software/nuca-mobile/assets/32801760/520b391a-10a3-4a27-b612-52fb63a15788

File: screens/ReportGenerationScreen.tsx (hunk 5 above)
Comment (sutgeorge) [reply screens/ReportGenerationScreen.tsx]: I managed to fix this issue anyways &rarr; https://github.com/crafting-software/nuca-mobile/pull/75#discussion_r1629088627

**Comment by sutgeorge** [comment]
https://github.com/crafting-software/nuca-mobile/pull/75#discussion_r1629088627

File: components/Footer.tsx (hunk 6 above)
Comment (sutgeorge) [reply components/Footer.tsx]: The banner still looks weird on iPhone 4 (although to be fair, that's a very old device and it is [discontinued](https://en.wikipedia.org/wiki/IPhone_4) anyways)

Comment (salajandan) [reply components/Footer.tsx]: @sutgeorge just a mention: this is just the *web* preview of how the app looks like on those devices, given the viewport size.
The iOS app could look slightly different on those devices, but i highly doubt it's installable on them.

Comment (sutgeorge) [reply components/Footer.tsx]: It is definitely not installable, especially on iPhone 4, since it is discontinued. Also, other devices such as Blackberry phones are probably not even used in 2024 either, due to the same reason previously mentioned (discontinuation/deprecation).

Comment (sutgeorge) [reply components/Footer.tsx]: Still, it is worth trying out multiple devices for the sake of ensuring that the app is responsive on several screen sizes.

File: components/Footer.tsx [hunk 7]
```diff
@@ -4,21 +4,26 @@ import catLady3 from '../assets/cat-lady3.png';
...
 const getFooterStyles = (_theme: NucaCustomTheme) =>
   StyleSheet.create({
     imageView: {
       marginTop: 32,
+      flex: 1,
     },
     image: {
+      alignSelf: 'center',
+      position: 'absolute',
+      bottom: isSmallMobileScreen() ? '-20%' : 0,
```
Comment (salajandan) [inline components/Footer.tsx:21]: can we avoid using negative offsets?

File: components/NucaFormButton.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      fontFamily: 'Nunito_700Bold',
+    },
+  });
+
+type NucaFormButtonProps = {
+  label: string;
+  iconName: string;
+  backgroundColor: string;
+  labelColor: string;
+  onPress: () => any;
```
Comment (salajandan) [inline components/NucaFormButton.tsx:34]: shouldn't this be `void` instead of `any` ?

**Review by salajandan — APPROVED** [review]
after comments

File: components/Footer.tsx (hunk 7 above)
Comment (sutgeorge) [reply components/Footer.tsx:21]: Given that we have this issue when using only `bottom: 0`, I'm afraid that it is not possible (using `top` doesn't make sense, as we attempt to position the banner relative to the bottom of the screen, rather than the top):

spatiu


The problem is that the banner is positioned a bit too far away from the bottom of the screen, one some devices.
